DOCSP-42335: Change streams (#69)
mongodb/docs-scala: source/read/change-streams.txt

and assign the following values to your ``database`` and ``collection`` variables:

.. literalinclude:: /includes/read/change-streams.scala
    :language: scala
    :dedent:
    :start-after: start-db-coll
    :end-before: end-db-coll

.. tip::

   To learn how to create a free MongoDB Atlas cluster and load the sample datasets,
   see the :atlas:`Get Started with Atlas </getting-started>` guide.

Some examples use instances of the ``LatchedObserver`` class to handle change
stream events. This class is a custom observer that prints change stream events
and continues monitoring data changes until the stream completes or generates
an error. To use the ``LatchedObserver`` class, paste the following code into
your application file:

.. literalinclude:: /includes/read/change-streams.scala
    :language: scala
    :dedent:
    :start-after: start-latched-observer
    :end-before: end-latched-observer

Open a Change Stream
--------------------

To open a change stream, call the ``watch()`` method. The instance on which you
call the ``watch()`` method determines the scope of events that the change
stream monitors. You can call the ``watch()`` method on instances of the following
classes:

> - ``MongoClient``: Monitor all changes in the MongoDB deployment

Reviewer: Q: does this mean all changes across all collections in all databases? if so, probably worth being explicit about that. i'm thinking there are a lot of config changes you could make to a deployment, like sharding and whatnot

Author: Yeah changes to all non-system collections across all databases, according to the Server docs. I'll reword to make that clear

- ``MongoDatabase``: Monitor changes in all collections in the database

> - ``MongoCollection``: Monitor changes in the collection

Reviewer: ```suggestion
- ``MongoCollection``: Monitor changes in one collection
```

The following example calls the ``watch()`` method to open a change stream on the
``restaurants`` collection. The code creates a ``LatchedObserver`` instance to
receive and output changes as they occur:

.. literalinclude:: /includes/read/change-streams.scala
   :start-after: start-open-change-stream
   :end-before: end-open-change-stream
   :language: scala
   :dedent:

> To begin watching for changes, run the preceding code. Then, in a separate
> shell, modify the ``restaurants`` collection. The following example updates
> a document that has a ``name`` field value of ``"Blarney Castle"``:

Reviewer: ```suggestion
To begin watching for changes, run the preceding code. Then, in a separate
shell, run the following code, which updates
a document that has a ``name`` field value of ``"Blarney Castle"``:
```

.. _scala-change-stream-update:

.. literalinclude:: /includes/read/change-streams.scala
   :start-after: start-update-for-change-stream
   :end-before: end-update-for-change-stream
   :language: scala
   :dedent:

> When you update the collection, the change stream application prints the change

Reviewer: ```suggestion
When you run the preceding code to update the collection, the change stream application prints the change
```

as it occurs. The printed change event resembles the following output:

.. code-block:: none
   :copyable: false

   ChangeStreamDocument{ operationType=update, resumeToken={"_data": "..."},
   namespace=sample_restaurants.restaurants, destinationNamespace=null, 
   fullDocument=null, fullDocumentBeforeChange=null, documentKey={"_id": {...}},
   clusterTime=Timestamp{...}, updateDescription=UpdateDescription{removedFields=[],
   updatedFields={"cuisine": "Irish"}, truncatedArrays=[], disambiguatedPaths=null},
   txnNumber=null, lsid=null, splitEvent=null, wallTime=BsonDateTime{...}}

Modify the Change Stream Output
-------------------------------

To modify the change stream output, you can pass a list of pipeline stages as a
parameter to the ``watch()`` method. You can include the following stages in the
list:

> - ``$addFields`` or ``$set``: Adds new fields to documents

Reviewer: S: somewhere in this section, include a link to Server docs that describe these stages in more detail

https://www.mongodb.com/docs/manual/reference/operator/aggregation-pipeline/

- ``$match``: Filters the documents
- ``$project``: Projects a subset of the document fields
- ``$replaceWith`` or ``$replaceRoot``: Replaces the input document with the
  specified document
- ``$redact``: Restricts the contents of the documents
- ``$unset``: Removes fields from documents

The {+driver-short+} provides the ``Aggregates`` class, which includes helper methods
for building the preceding pipeline stages. To learn more about ``Aggregates``, see
the `API documentation <{+api+}/org/mongodb/scala/model/Aggregates$.html>`__.

> The following example passes a pipeline to ``watch()`` that uses the ``Aggregates.filter()``
> method to build the ``$match`` stage. This instructs the ``watch()`` method to output
> events only when update operations occur:

Reviewer: ```suggestion
The following example creates a pipeline that uses the ``Aggregates.filter()``
method to build the ``$match`` stage. It then passes this pipeline to the
``watch()`` method. This instructs the ``watch()`` method to output
events only when update operations occur.
```

.. literalinclude:: /includes/read/change-streams.scala
   :start-after: start-change-stream-pipeline
   :end-before: end-change-stream-pipeline
   :language: scala
   :dedent:

Modify watch() Behavior
-----------------------

You can modify the behavior of the ``watch()`` method by chaining
methods provided by the ``ChangeStreamObservable`` class. The following
table describes some of these methods:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Method
     - Description

   * - ``fullDocument()``
     - | Specifies whether to show the full document after the change, rather
         than showing only the changes made to the document. To learn more about
         this option, see the :ref:`scala-change-stream-pre-post-image` section of this
         guide.

   * - ``fullDocumentBeforeChange()``
     - | Specifies whether to show the full document as it was before the change, rather
         than showing only the changes made to the document. To learn more about
         this option, see :ref:`scala-change-stream-pre-post-image`.

   * - ``comment()``
     - | Attaches a comment to the operation.

   * - ``startAtOperationTime()``

>      - | Instructs the change stream to only provide changes that occurred at or after

Reviewer: ```suggestion
     - | Instructs the change stream to provide only changes that occurred at or after
```

         the specified timestamp.

   * - ``collation()``
     - | Sets the collation to use for the change stream cursor.

For a full list of ``watch()`` options, see `ChangeStreamObservable
<{+api+}/org/mongodb/scala/ChangeStreamObservable.html>`__ in the API
documentation.

.. _scala-change-stream-pre-post-image:

Include Pre-Images and Post-Images
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. important::

   You can enable pre-images and post-images on collections only if your

>    deployment uses MongoDB v6.0 or later.

Reviewer: ```suggestion
   deployment uses {+mdb-server+} v6.0 or later.
```

By default, when you perform an operation on a collection, the

> corresponding change event includes only the delta of the fields

Reviewer: S: some readers might not know "delta" in this context. could either reword or link to something like this:

https://en.wikipedia.org/wiki/Delta_encoding

modified by that operation. To see the full document before or after a
change, chain the ``fullDocumentBeforeChange()`` or the ``fullDocument()``

> methods to the ``watch()`` method.

Reviewer: S: I think I would cut this whole sentence, and just present the methods sequentially below. It's tricky to correlate the methods in the second clause back to their purposes in the first.

If you make this change, also modify the pre-image/post-image paragraphs to mention chaining the methods to `watch()`

post-image in the change stream event, pass one of the following values to the
``fullDocument()`` method:

- ``FullDocument.UPDATE_LOOKUP``: The change event includes a
  copy of the entire changed document from some time after the change.
- ``FullDocument.WHEN_AVAILABLE``: The change event includes
  a post-image of the modified document for change events. If the post-image is not
  available, this change event field has a ``null`` value.
- ``FullDocument.REQUIRED``: The change event includes a post-image
  of the modified document for change events. If the post-image is not available, the
  server raises an error.

The following example calls the ``watch()`` method on a collection and includes the post-image
of updated documents by chaining the ``fullDocument()`` method:

.. literalinclude:: /includes/read/change-streams.scala
   :start-after: start-change-stream-post-image
   :end-before: end-change-stream-post-image
   :language: scala
   :dedent:

With the change stream application running in a separate shell, updating a
document in the ``restaurants`` collection by using the :ref:`preceding update

> example <scala-change-stream-update>` prints a change event resembling the following

Reviewer: ```suggestion
example <scala-change-stream-update>` prints a change event that resembles the following
```

         guide.

   * - ``fullDocumentBeforeChange()``
     - | Specifies whether to show the full document as it was before the change, rather
         than showing only the changes made to the document. To learn more about
         this option, see :ref:`scala-change-stream-pre-post-image`.

   * - ``comment()``
     - | Attaches a comment to the operation.

   * - ``startAtOperationTime()``
     - | Instructs the change stream to provide only changes that occurred at or after
         the specified timestamp.

   * - ``collation()``
     - | Sets the collation to use for the change stream cursor.

For a full list of ``watch()`` options, see `ChangeStreamObservable
<{+api+}/org/mongodb/scala/ChangeStreamObservable.html>`__ in the API
documentation.

.. _scala-change-stream-pre-post-image:

Include Pre-Images and Post-Images
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. important::

   You can enable pre-images and post-images on collections only if your
   deployment uses {+mdb-server+} v6.0 or later.

By default, when you perform an operation on a collection, the
corresponding change event includes only the modified fields and
their values before and after the operation. 

> You can instruct the ``watch()`` method to return the document's **pre-image**, or the

Reviewer: Q: is this returned in addition to the modified fields, or instead of them? probably worth saying. 

also applies to post-image

stream monitors. You can call the ``watch()`` method on instances of the following
classes:

- ``MongoClient``: Monitors changes to all non-system collections across all databases

>   in a deployment

Reviewer: S: give quick examples of "system" collections

> - ``MongoDatabase``: Monitors changes in all collections in one database
> - ``MongoCollection``: Monitors changes in one collection

Reviewer: ```suggestion
- ``MongoDatabase``: Monitors changes to all collections in one database
- ``MongoCollection``: Monitors changes to one collection
```

.. important::

   You can enable pre-images and post-images on collections only if your
   deployment uses {+mdb-server+} v6.0 or later.

By default, when you perform an operation on a collection, the
corresponding change event includes only the modified fields and
their values before and after the operation. 

> You can instruct the ``watch()`` method to return the document's **pre-image**, or the

Reviewer: S: "or" might be misleading here

```suggestion
You can instruct the ``watch()`` method to return the document's **pre-image**, the
```

> full version of a document *before* a change. To include the pre-image in

Reviewer: ```suggestion
full version of the document *before* changes. To include the pre-image in
```

  of the modified document for change events. If the pre-image is not available, the
  server raises an error.  

You can also instruct the ``watch()`` method to return the document's **post-image**,

> or the full version of a document *after* a change. To include the

Reviewer: ```suggestion
the full version of a document *after* changes. To include the
```

> or the full version of a document *after* a change. To include the

Reviewer: ```suggestion
or the full version of the document *after* a change. To include the
```
